### PR TITLE
Hello JDK 17

### LINKED ISSUE #31: Support for Java 17
Currently some parts of the code are not compatible with next LTS release which is Java 17(Hello Reflection API). This issue is created for tracking and to inform everyone that this issue is acknowledged and being worked on.

**The problem:** 
Currently Gson library that we are using for JSON serialization, by default uses Reflection API somewhere underneath to serialize JSON, but since in 2021 Java decided that strong encapsulation now important it limited access to some classes that Reflection API previously could access, hence GSON was working fine with some default configuration.
Good news tho, this problem can be solved with creating custom  [TypeAdapter](https://www.javadoc.io/doc/com.google.code.gson/gson/2.8.8/com.google.gson/com/google/gson/TypeAdapter.html). Although, that will involve some internal code changes, but nothing should break from end user perspective(hopefully).

In addition, some dependencies need to be updated as well, to ensure compatibility with Java 17.

**Closing criterias:**
*  Dependencies are updated to the latest to ensure compatibility with Java 17.
* Custom `TypeAdapters` created for models, that need them(Luckily, not all of them require custom adapters).
* Tests can be run without issues with Java 8, 11, 17

### PR DESCRIPTION
**Pull request by mihhail-m**
closes #31 

**TLDR:**
* `TestSerializer/Deserialzer` replaced by `TestTypeAdapter`. Difference between the two is that `TypeAdapter` operates on lower level then previously used `JsonSerializer/Deserializer`, hence allowing us working around JDK 17 changes. Also, less files to maintain. 
_Note for the future_: maybe replace all `Serializer/Deserializer`s with `TypeAdapter`? 
* Updated dependencies to comfort JDK 17.
* `Test` had to be marked with `ModelParams` interface in order to be able to pass it further down to `.request()` method. This is a lazy workaround to some issues caused by moving to `TypeAdapter` implementation. 
* Removed one unnecessary test. Updated another.
  
**P.S**: I think it's time to make fresh release.

### CODE REVIEW

File: src/main/java/com/loadero/http/TestTypeAdapter.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    public void write(JsonWriter jsonWriter, Test test) throws IOException {
+        if (Objects.isNull(test)) {
+            jsonWriter.nullValue();
+        } else {
+            jsonWriter.beginObject();
+
+            jsonWriter.name("created").value(test.getCreated());
+            jsonWriter.name("updated").value(test.getUpdated());
+            jsonWriter.name("id").value(test.getId());
+            jsonWriter.name("project_id").value(test.getProjectId());
```
Comment (ajpetersons) [inline src/main/java/com/loadero/http/TestTypeAdapter.java:96]: these 4 values cannot be sent to Loadero API as they must be empty fields (readonly)

File: src/main/java/com/loadero/http/TestTypeAdapter.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            jsonWriter.name("name").value(test.getName());
+            jsonWriter.name("script_file_id").value(test.getId());
```
Comment (ajpetersons) [inline src/main/java/com/loadero/http/TestTypeAdapter.java:98]: this value cannot be sent to Loadero API as they must be empty fields (readonly). instead we need a `script` field with text value of the test script

**Review by ajpetersons — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 3]
```diff
@@ -5,7 +5,7 @@
     <modelVersion>4.0.0</modelVersion>
     <groupId>com.loadero</groupId>
     <artifactId>loadero-java</artifactId>
-    <version>1.1.2</version>
```
Comment (ajpetersons) [inline pom.xml:8]: 1.1.2 isn't yet released, so we can revert this change

File: src/main/java/com/loadero/http/TestRunTypeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        int participantCount = testRunMap.get(PARTICIPANT_COUNT) == null ? 0 :
+            (int) testRunMap.get(PARTICIPANT_COUNT);
+        String processingStarted = testRunMap.get(PROCESSING_STARTED) == null ? ""
+            : (String) testRunMap.get(PROCESSING_STARTED);
+        String processingFinished = testRunMap.get(PROCESSING_FINISHED) == null ? ""
+            : (String) testRunMap.get(PROCESSING_FINISHED);
+        String executionStarted = testRunMap.get(EXECUTION_STARTED) == null ? ""
+            : (String) testRunMap.get(EXECUTION_STARTED);
+        String executionFinished = testRunMap.get(EXECUTION_FINISHED) == null ? ""
+            : (String) testRunMap.get(EXECUTION_FINISHED);
```
Comment (theRealAlpaca) [inline src/main/java/com/loadero/http/TestRunTypeAdapter.java:141]: You can fit some of these in one line  and the one that do not fit 120 chars, wrap like this:
```java
        int participantCount = testRunMap.get(PARTICIPANT_COUNT) == null ? 0 : (int) testRunMap.get(PARTICIPANT_COUNT);
        String processingStarted = testRunMap.get(PROCESSING_STARTED) == null
                ? ""
                : (String) testRunMap.get(PROCESSING_STARTED);
````

File: src/main/java/com/loadero/http/TestTypeAdapter.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            jsonWriter.nullValue();
+        } else {
+            jsonWriter.beginObject();
+
+            jsonWriter.name("name").value(test.getName());
+            jsonWriter.name("script").value(test.getScript());
+            jsonWriter.name("start_interval").value(test.getStartInterval().toString());
+            jsonWriter.name("participant_timeout").value(test.getParticipantTimeout().toString());
+            jsonWriter.name("mode").value(test.getMode().toString());
+            jsonWriter.name("increment_strategy").value(test.getIncrementStrategy().toString());
```
Comment (theRealAlpaca) [inline src/main/java/com/loadero/http/TestTypeAdapter.java:98]: You can use constants declared in the beginning of the class here

**Review by theRealAlpaca — CHANGES_REQUESTED** [review]

File: src/main/java/com/loadero/http/TestTypeAdapter.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        TestParams params = paramsBuilder.build();
+
+        return new Test(params);
+    }
+
+    @Override
+    public void write(JsonWriter jsonWriter, Test test) throws IOException {
+        if (Objects.isNull(test)) {
+            jsonWriter.nullValue();
+        } else {
```
[the remark this answers is not in the document]
Comment (mihhail-m) [reply src/main/java/com/loadero/http/TestTypeAdapter.java:90]: What do you mean by "early return"? This is a `void` function so nothing is being returned.
`test == null` it is. Just to please y'all Go devs

File: src/main/java/com/loadero/http/TestTypeAdapter.java (hunk 5 above)
Comment (mihhail-m) [reply src/main/java/com/loadero/http/TestTypeAdapter.java:98]: good point. Didn't notice that. Thank you.

File: src/main/java/com/loadero/http/TestTypeAdapter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (theRealAlpaca) [reply src/main/java/com/loadero/http/TestTypeAdapter.java:90]: 😄 I am not pushing `test == null` approach. To me it just seems easier on eyes and requires one less import.

Regarding early returns I meant this
```java
if (test == null) {
            jsonWriter.nullValue();
            return;
}

// stuff written in else block
```

[the remark this answers is not in the document]
Comment (mihhail-m) [reply src/main/java/com/loadero/http/TestTypeAdapter.java:90]: chill, i was joking. I am all in myself with less imports especially in Java.
Not sure if that `return` statement is needed? As this really should terminate right after `jsonWriter.nullValue();` anyway. But can add it, I guess.

[the remark this answers is not in the document]
Comment (mihhail-m) [reply src/main/java/com/loadero/http/TestTypeAdapter.java:90]: yeah, even IDE screaming at me  and telling, that this `return` statement in the `void` function is unnecessary.

[the remark this answers is not in the document]
Comment (theRealAlpaca) [reply src/main/java/com/loadero/http/TestTypeAdapter.java]: Does it? To me IntelliJ reports warnings only if there is `else` block. This looks to be working fine:
```java
public void write(JsonWriter jsonWriter, Test test) throws IOException {
    if (test == null) {
        jsonWriter.nullValue();
        return;
    }

    jsonWriter.beginObject();

    jsonWriter.name("name").value(test.getName());
    jsonWriter.name("script").value(test.getScript());
    jsonWriter.name("start_interval").value(test.getStartInterval().toString());
    jsonWriter.name("participant_timeout").value(test.getParticipantTimeout().toString());
    jsonWriter.name("mode").value(test.getMode().toString());
    jsonWriter.name("increment_strategy").value(test.getIncrementStrategy().toString());

    jsonWriter.endObject();
}
```

[the remark this answers is not in the document]
Comment (mihhail-m) [reply src/main/java/com/loadero/http/TestTypeAdapter.java]: alright, lets fight over that `return` statement. 

I guess, my IntelliJ just smarter 🙄 
Screenshot 2022-01-05 at 11 30 56

**Review by theRealAlpaca — APPROVED** [review]
